api/swagger.yaml

```diff
@@ -8876,13 +8876,19 @@ paths:
                   SubnetSize specifies the subnet size of the networks created from the default subnet pool
                 type: "integer"
                 format: "uint32"
+              VxlanUDPPort:
```

Comment:
`VxlanUDPPort` is a bit verbose here, and very much tied to the implementation; perhaps some other naming would be better.

*However* - if I'm looking at this (and understand correctly), this configures the port to used for the overlay data path. We already have `DataPathAddr`, which defines the address to use. `ListenAddr` and `AdvertiseAddr` both accept an ip/interface and optionally a port. Why don't we use the same approach for this one? So;

Change `DataPathAddr` to accept the format: `<ip|interface>[:<port>]`

- if only `<ip|interface>` is specified, use the default port (4789)
- if both `<ip|interface>` and `<port>` are specified, use the given address and port
- if none is specified, use `ListenAddr` as address, and `4789` as port.

That change would be backward-compatible, because the port is optional


I _think_ this approach actually came up in a discussion when the `DataPathAddr` was added (https://github.com/moby/moby/pull/32717) I can't find that discussion though, so possibly it was offline (or on Slack). @fcrisciani may recall

Reply:
We would like to use single VXLAN port across all the node in the same cluster.  Datapathaddr can be configured via swarm init and  as well swarm join.  Issue here is user will have to remember and pass same port number for each swarm join. Thats going to be error prone. We kind of discussed this option and moved away from it.
I wanted to make sure we discuss  design related issues before starting feature development  and thats the main reason I had written proposal https://github.com/moby/moby/issues/38030 ( before working on this feature )to make sure we discuss any design specific issue before implementation. 
Pls let me know if you still have concern with this approach. 

Regrading VxlanUDPPort name : I wanted to convey this port is used only for vxlanport ( on UDP). Initial name was suggested like overlayvxlanport or something like that. But I thought this is simple :). Let me know, I can try to come up with some other name.

Reply:
> Datapathaddr can be configured via swarm init and as well swarm join. Issue here is user will have to remember and pass same port number for each swarm join. Thats going to be error prone.

Let me try to understand; 

#### Old situation (current release)

- `docker swarm init` initializes the swarm, and uses port `4789` (hard-coded)
- `docker swarm join` joins a node to the swarm, and also uses port `4789` (hard-coded)


### Current flow in this PR

- Run `docker swarm init --vxlan-udp-port=1234` to initialize it as a manager
- On another node, run `docker swarm join <manager-ip>` to join the node to the swarm
- The node somehow gets the port number from the manager, and sets `vxlan-udp-port` to the same port as the manager

or

- On another node, run `docker swarm join --vxlan-udp-port=4567 <manager-ip>` to join the node to the swarm, but using a different port-number than is used by the manager


Question:

- How is the port communicated from the manager to the node that is joining? Or will this be printed as part of the ouput of `docker swarm init`? ("To add a worker to this swarm, run the following command:")? 
- I wonder if we can make the assumption that all nodes in the swarm are expected to be using the same port (I can foresee situations where a different port number is needed for _some_ nodes and/or data-centers, but possibly not for _all_ the nodes)

### Alternative proposal

So, thinking about my proposal;

- Run `docker swarm init --data-path-addr=1.2.3.4:4321` to initialize a swarm
- Run `docker swarm join` to join a node
  - should this use the default (`4789`) port?
  - or, should this use the same port as the manager uses (`4321`)?
  - note that the manager can still communicate the port-number to the node (as it _does_ have its portnumber set)
- Run `docker swarm join --data-path-addr=2.3.4.5:1234` to join the node to the swarm, but using a different port-number than is used by the manager

> Or will this be printed as part of the ouput of `docker swarm init`? ("To add a worker to this swarm, run the following command:")? 

If that is the case, we can consider making the `<ip|interface>` optional;

`--data-path-addr` | data-path-address | vxlan port
---|---|---
(not set) | `ListenAddr` (default) | `4789` (default)
`1.2.3.4` | `1.2.3.4` |  `4789` (default)
`eth1` | IP-addres of `eth1` |  `4789` (default)
`1.2.3.4:4321` | `1.2.3.4` |  `4321`
`eth1:4321` | IP-addres of `eth1` |  `4321`
`:4321` | `ListenAddr` (default) |  `4321`


The above is the approach that's used for the `-H` option of `dockerd`;

```
dockerd -H 172.18.0.1
INFO[2018-11-09T15:03:37.583248272Z] API listen on 172.18.0.1:2375 

dockerd -H 172.18.0.1:4321
INFO[2018-11-09T15:03:37.583248272Z] API listen on 172.18.0.1:4321 

dockerd -H :4321
INFO[2018-11-09T15:00:19.352779504Z] API listen on 127.0.0.1:4321   
```

### Naming

> Regrading VxlanUDPPort name : I wanted to convey this port is used only for vxlanport ( on UDP). Initial name was suggested like overlayvxlanport or something like that. But I thought this is simple :). Let me know, I can try to come up with some other name.

I think that's something to be communicated through the _documentation_ (both for the CLI, and in the description of the Swagger) and (if room permits) the flag description. 

For the flag description, this could be something like one of these;

```
Port to use for data path traffic (UDP)
Port (UDP) to use for data path traffic
UDP port to use for data path traffic
```

Given that this relates to the data-path,  my preference would be to name both the same / similar:

- `DataPathPort` for the API
- `--data-path-port` for the CLI

That way, it's clear what this option does, without having to understand what VxLAN is used for (and/or leaking that implementation detail into the UX). It will also show both options together in the CLI usage output;

```
Usage:	docker swarm join [OPTIONS] HOST:PORT

Join a swarm as a node and/or manager

Options:
      --advertise-addr string   Advertised address (format: <ip|interface>[:port])
      --availability string     Availability of the node ("active"|"pause"|"drain") (default "active")
      --data-path-addr string   Address or interface to use for data path traffic (format: <ip|interface>)
      --data-path-port int      Port to use for data path traffic (UDP) (default 4789)
      --listen-addr node-addr   Listen address (format: <ip|interface>[:port]) (default 0.0.0.0:2377)
      --token string            Token for entry into the swarm
```

Reply:
I had offline discussion with @thaJeztah and summary is below:

data-path-addr is configured per node ( manager or worker). So we added option in swarm init & swarm join.
vxlanport - is configured per cluster. ( Main  use case for this feature  is what if port "4789"  is blocked.  In general customers prefer to stick to single port because if they use different port for each node, then trying to open/expose  port for any service - they need to remember all the ports that have been used in each and every node in the cluster. It will create unnecessary headache  )

vxlan port number will be communicated via node object information from leader to worker node after they join the cluster.

 Naming:
=======

I still prefer to keep it vxlan-udp-port at the end of the day this what it represents.   But I can change it to --data-path-port instead of vxlanport.
Will go ahead and modify the name.